GitUI/UserControls/RevisionGrid/RevisionGridControl.cs

```diff
@@ -914,40 +916,61 @@ public void PerformRefreshRevisions(Func<RefsFilter, IReadOnlyList<IGitRef>> get
                 _gridView.SelectionChanged += OnGridViewSelectionChanged;
                 _gridView.ResumeLayout();
 
-                string pathFilter = BuildPathFilter(_filterInfo.PathFilter);
-                _revisionReader.Execute(
-                    Module,
-                    refs,
-                    revisions,
-                    _filterInfo.CommitsLimit,
-                    _filterInfo.RefFilterOptions,
-                    _filterInfo.IsShowFilteredBranchesChecked ? _filterInfo.BranchFilter : string.Empty,
-                    _filterInfo.GetRevisionFilter(),
-                    pathFilter);
-
-                if (_initialLoad)
-                {
-                    _selectionTimer.Enabled = false;
-                    _selectionTimer.Stop();
-                    _selectionTimer.Enabled = true;
-                    _selectionTimer.Start();
-
-                    _initialLoad = false;
-                }
+                // Add back and show the spinner control, which was removed by SetPage call
+                Controls.Add(_loadingControlSync);
+                ShowLoading();
 
-                _superprojectCurrentCheckout = null;
                 ThreadHelper.JoinableTaskFactory.RunAsync(async () =>
                 {
-                    var scc = await GetSuperprojectCheckoutAsync(capturedModule, noLocks: true);
-                    await this.SwitchToMainThreadAsync();
-
-                    if (_superprojectCurrentCheckout != scc)
+                    cancellationToken.ThrowIfCancellationRequested();
+                    await TaskScheduler.Default;
+
+                    // Find all ambiguous refs (including stash, notes etc)
+                    // NB: This can be EXTREMELY slow operation for large repos.
```

Comment:
`_ambiguousRefs` aren't used in this method at all, and is a placeholder for some other on-demand functionality. Since it isn't actually needed, wrapping it in a Lazy avoids doing unnecessary work.

The VS repo has ~180K tags, which make GetRefs call very expensive. I guess the Linux repo doesn't suffer from the same insanity. In my test scenarios I disabled the Tags node in left panel, and made the revision grid loading (that includes loading refs and revisions) async. This makes the app appear significantly more responsive.

Reply:
Slightly OT: How slow was this at startup in post 3.5 before #9734/#9735 when GefRefs was running like 9 times?
And in 3.5 this is like 5 times (plus running 4 times when opening Dashboard?
(one time in master)
Getting ahead/behind is done in two steps and with many refs this would be improved by sharing similar to GetRefs().